### DISCUSSION TITLE
A concerning trend?

### DISCUSSION THREAD

**Discussion by pressRtowin** [TrianguloY/URLCheck#316] [General]
I'm just going to copy something I wrote over in the LinkSheet repo (https://github.com/1fexd/LinkSheet/issues/321) here because it's relevant to URLCheck too and I'm hoping to get some discussion going about what (if anything) can be done about this.

> I don't know if you've noticed, but a few months back, Discord stated pushing out what I consider to be a rather concerning update. All links opened from Discord now completely ignore your system "Opening Links" settings. There seems to be a list of apps directly programed to open for any of their associated domains. If you attempt to open ANY link to one of those domains, it will force it to open in its respective app, if that app is installed. This means, for example, that you can no longer have YouTube Revanced installed while also leaving the YouTube app enabled so videos in the Play Store still work. If you want YouTube links from Discord to open in Revanced, you MUST disable the stock YouTube app. Even worse is that even after disabling the stock app, you still get no control over which alternate app (that can handle YouTube links) Discord will pick. It will always pick the next one in its list (not sure how it orders them), so if the app you want it to use is 4th in its list for that domain, you'll have to disable all 3 apps above it. Again, this behavior happens _regardless_ of the "Opening Links" settings for each of those apps.
> 
> For domains that aren't associated with an app on the list, attempting to open a link from Discord _will_ respect your default browser setting, if you have one set. If you don't have one set, you no longer get a system dialog to pick which app you want to use. The link will _always_ open in Chrome.
> 
> So this means that in both cases above, you'll now _never_ see the system app picker when opening links from Discord. Furthermore, it means that as more and more domains get white listed (and associated with specific apps), even setting app picker replacements like LinkSheet as the default browser will become less and less useful.
> 
> I know that was a long and seemingly off-topic rant, but the reason I'm bringing all this up, I guess, is I'm wondering if you've run into any other apps doing this recently, and if you know _any_ way around this. I fear that this may turn into a pattern sooner than later. Do apps like LinkSheet have any way to work around this? Or are we at the mercy of app developers not doing whatever the hell it is that Discord is doing? I even tried LinkSheetCompat, as its description (about Chrome searching for YouTube explicitly) seemed exactly like what Discord is doing, but apparently what Discord is doing is even more pervasive than Chrome, and LinkSheetCompat was never opened, even after disabling stock YouTube.
> 
> Any thoughts/ideas?

**Comment by pressRtowin** [comment]
Worth noting that URLCheck actually offers the closest thing to what you might call a "solution" to this so far.

I can long press on a link in Discord, then copy it to the clipboard. Then I can use my URLCheck widget in the popup widget panel to open URLCheck with the URL loaded. A bit more cumbersome than just tapping on a link, but better than nothing.

**Comment by TrianguloY** [comment]
Unfortunately I have not experienced the described issue yet, although I don't think I have any of those "blacklisted" apps. All links I use are opened with URLCheck, except YouTube which is opened on the _correct_ app (because I have the other disabled).

I know that Android has a concept of "ownership" of links, and an app can register a link to be always opened by it. This has always been possible to be disabled by the user from the android settings though.

In fact, even if discord is choosing specifically which app to open, it depends on the apps that Android reports are available. You should theoretically be able to disable an app from opening links, and this should make discord not "see" the app and thus not opening it. If this is not working...maybe google (android maker) has some blame too...

But I cannot give a correct response, only my opinion. What I'll definitely do is to keep an eye on this behavior, Thanks for reporting! 

~~Btw: if you want you can link here the other post, it may be useful in case there are more discussion there!~~ Edit: oops, it was already! Sorry.

**Reply by pressRtowin** [reply]
>I know that Android has a concept of "ownership" of links, and an app can register a link to be always opened by it. This has always been possible to be disabled by the user from the android settings though.
>
> In fact, even if discord is choosing specifically which app to open, it depends on the apps that Android reports are available. You should theoretically be able to disable an app from opening links, and this should make discord not "see" the app and thus not opening it. If this is not working...maybe google (android maker) has some blame too...

That's the thing. Discord is completely ignoring this setting now.

It appears that what's happening is that it has a way to generate a list of all apps that are able to handle one of its "white listed" links. The first is always the first party app, then other compatible apps in some set order. It will then go down this list any only skip to the next entry if the current one it's looking for is not installed or _disabled_. The "Opening Links" system setting has _no_ effect on its behavior. I keep the first party YouTube app enabled so embedded YouTube players (such as in the Play Store) will continue to work. Discord will now _only_ open YouTube links with that app. If I disable that app and enable opening links on Revanced Extended but not Revanced, it doesn't matter, as apparently Revanced is higher on its list than Revanced Extended is. Now every app opens in Revanced. Only after disabling both Revanced and YouTube can I get it to open in Extended (assuming no other apps that can handle YouTube links are higher in its list.

Another white listed app I've discovered is Twitter.

Even with other domains, yeah apps like URLCheck still work for now (until more domains get white listed), but only if you have it set as the default browser (as I'm sure you do). We've lost access to the system browser selection menu _entirely_ (when using discord anyway).

There's some good discussion happening about this over in the issue I linked that you should check out. But it's sounding like the only way to truly address this behavior will be to patch offending apps with LSPatch or something similar.

**Reply by TrianguloY** [reply]
Yeah, I was reading the discussion.

It's like discord is testing the apps on its hardcoded list, so yeah only by disabling or uninstalling them it will pick the next.

But that Revanced situation...is concerning. I highly doubt that discord has revanced in their list, so it means that it is getting the list of apps directly from android. This is what I think needs to be investigated, because this list should be able to be "edited" by the user, and if it isn't...then android is the one to blame here.

There is also the possibility that they are doing both, first check the internal list, then check the android one, but that still doesn't explain why you need to disable the first revanced in order to show the second...

**Reply by pressRtowin** [reply]
Chrome also does a "diet" version of what Discord is doing, but LinkSheet's dev was able to work around it with https://github.com/1fexd/LinkSheetCompat. However, Discord works around even that and explicitly calls on apps to open directly somehow.

**Reply by pressRtowin** [reply]
> I highly doubt that discord has revanced in their list, so it means that it is getting the list of apps directly from android. 

The only thing that makes sense to me is that they're querying the list of compatible apps via some other method that gives them a full list, rather than just what you have enabled. Once that's done, the first party app is brought to the front of the list, the rest are either left as-is or they're sorted some other way, and then opening is prioritized by the final order of that list, no matter what. So no app further down on the list will be opened unless _every_ app above it is disabled.

**Reply by pressRtowin** [reply]
Here's a video of the behavior demonstrated:

https://github.com/TrianguloY/UrlChecker/assets/25999169/158e6eaf-c0c4-45e7-82c2-2a5ea0fe8195

**Reply by PabloOQ** [reply]
If Discord decides to choose the app they send the links to I don't think there is any solution in URLCheck. The only alternatives (at least that I can think of) to fix this is are:
- Spoofing the list of apps, so that discord has no information on which apps are installed. To be clear, I have no idea if this works, but maybe this can be easily done without patches/root by using the work profile, I think it sandboxes apps, but again, I'm not certain.
- Patching the app so it asks nicely to android to handle it instead.

On this topic, I have never reported this as it is too specific, but something similar happens with Twitter too. When taping URLs from Twitter without a defined default browser, Twitter will not wrap the URL with `t.co`, however, the moment an app is defined it will give you the `t.co` URL instead and use the first app of the app chooser (I think) until Twitter is closed. But I haven't looked into it, maybe it's "fixed" now (I haven't updated my twitter app in months) or it's just a bug on my phone.
Of course I don't expect this to be solved, but I wanted to show other instances of apps not asking android which app should be opened. We do the same in URLCheck, but that is the objective of the app.


https://github.com/TrianguloY/UrlChecker/assets/55493443/7e0f43d9-bd60-49c8-8112-daddc4028be2

_I close Twitter, open it again and go to the same tweet_

https://github.com/TrianguloY/UrlChecker/assets/55493443/3228b2f4-7e02-4d51-9ac7-f3499221e8f4

**Reply by pressRtowin** [reply]
> it will give you the `t.co` URL instead and use the first app of the app chooser (I think) until Twitter is closed

This behavior is wild. Also locking you out of other browsers once it switches to t.co. yet another reason why I hate all this verified domain nonsense.
